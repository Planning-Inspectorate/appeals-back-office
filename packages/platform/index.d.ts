import { HandlerFunction } from 'got';

declare module '@pins/platform' {
	export function createTtlHandler(): HandlerFunction;
	export function validatePostcode(value: string): boolean;
	export function validatePastDate(value: Date | string | number): boolean;
	export function validateFutureDate(value: Date | string | number): boolean;
	export function yesterday(): Date;
}

export interface LocalPlanningDepartment {
	FID: number;
	LPA21CD: string;
	LPA21NM: string;
	Co_terminous: 'Y' | 'N';
}
