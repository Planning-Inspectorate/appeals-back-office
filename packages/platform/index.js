export * from './http/ttl.js';
export * from './util/date.js';
export * from './validators/date.js';
export * from './validators/postcode.js';
