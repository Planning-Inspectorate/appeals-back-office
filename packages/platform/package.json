{
  "name": "@pins/platform",
  "version": "0.0.0",
  "type": "module",
  "main": "index.js",
  "private": true,
  "engines": {
    "node": ">=22.0.0 <= 22.14.0"
  },
  "scripts": {
    "clean": "rimraf .turbo node_modules dist",
    "lint:js": "npx eslint .",
    "lint:js:fix": "npx eslint . --fix",
    "test": "cross-env NODE_ENV=test NODE_OPTIONS=\"--experimental-vm-modules --expose-gc\" npx jest --logHeapUsage --ci",
    "test:cov": "cross-env NODE_ENV=test NODE_OPTIONS=\"--experimental-vm-modules --expose-gc\" npx jest --logHeapUsage --ci --coverage",
    "tscheck": "npx tsc -p jsconfig.json --maxNodeModuleJsDepth 0"
  },
  "dependencies": {
    "date-fns": "2.28.0",
    "dotenv": "16.0.3",
    "got": "^12.5.2",
    "html": "^1.0.0",
    "lodash-es": "^4.17.21",
    "node-html-parser": "^5.3.3",
    "randexp": "^0.5.3"
  },
  "devDependencies": {
    "@types/html": "^1.0.4",
    "@types/lodash-es": "^4.17.12",
    "cross-env": "^7.0.3",
    "jest": "^29.7.0",
    "type-fest": "^4.37.0"
  }
}
