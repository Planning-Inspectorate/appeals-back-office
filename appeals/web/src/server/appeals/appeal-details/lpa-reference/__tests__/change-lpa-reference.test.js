import { appealData } from '#testing/app/fixtures/referencedata.js';
import { createTestEnvironment } from '#testing/index.js';
import { parseHtml } from '@pins/platform';
import nock from 'nock';
import supertest from 'supertest';

const { app, teardown } = createTestEnvironment();
const request = supertest(app);
const baseUrl = '/appeals-service/appeal-details';

describe('change-lpa-reference', () => {
	beforeEach(() => {
		nock('http://test/')
			.get(`/appeals/${appealData.appealId}`)
			.reply(200, { ...appealData, lpaQuestionnaireId: null })
			.persist();
	});

	afterEach(teardown);

	describe('GET /change', () => {
		it('should render changeLpaReference page when loaded from appeal details', async () => {
			const appealId = appealData.appealId;
			const response = await request.get(`${baseUrl}/${appealId}/lpa-reference/change`);

			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(elementInnerHtml).toContain('What is the application reference number?</h1>');
			expect(elementInnerHtml).toContain('Continue</button>');
		});

		it('should render changeLpaReference page when loaded from appellant case', async () => {
			const appealId = appealData.appealId;
			const response = await request.get(
				`${baseUrl}/${appealId}/appellant-case/lpa-reference/change`
			);

			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(elementInnerHtml).toContain('What is the application reference number?</h1>');
			expect(elementInnerHtml).toContain('Continue</button>');
		});
	});

	describe('POST /change', () => {
		it('should re-render changeLpaReference page with an error when planningApplicationReference is empty', async () => {
			const appealId = appealData.appealId.toString();

			const invalidData = {
				planningApplicationReference: ''
			};
			const response = await request
				.post(`${baseUrl}/${appealId}/lpa-reference/change`)
				.send(invalidData);

			expect(response.statusCode).toBe(200);

			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(elementInnerHtml).toContain('What is the application reference number?</h1>');

			const unprettifiedErrorSummaryHtml = parseHtml(response.text, {
				rootElement: '.govuk-error-summary',
				skipPrettyPrint: true
			}).innerHTML;

			expect(unprettifiedErrorSummaryHtml).toContain('There is a problem</h2>');
			expect(unprettifiedErrorSummaryHtml).toContain('Enter the application reference number</a>');
		});

		it('should re-render changeLpaReference page with an error when planningApplicationReference is over 100 characters', async () => {
			const appealId = appealData.appealId.toString();

			const invalidData = {
				planningApplicationReference:
					'12345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901'
			};

			const response = await request
				.post(`${baseUrl}/${appealId}/lpa-reference/change`)
				.send(invalidData);

			expect(response.statusCode).toBe(200);

			const elementInnerHtml = parseHtml(response.text).innerHTML;

			expect(elementInnerHtml).toMatchSnapshot();
			expect(elementInnerHtml).toContain('What is the application reference number?</h1>');

			const unprettifiedErrorSummaryHtml = parseHtml(response.text, {
				rootElement: '.govuk-error-summary',
				skipPrettyPrint: true
			}).innerHTML;

			expect(unprettifiedErrorSummaryHtml).toContain('There is a problem</h2>');
			expect(unprettifiedErrorSummaryHtml).toContain(
				'Application reference number must be 100 characters or less</a>'
			);
		});

		it('should redirect to the appeal-details page when planningApplicationReference is valid and was opened in appeal details', async () => {
			const appealId = appealData.appealId.toString();
			nock('http://test/').patch(`/appeals/${appealId}`).reply(200, {
				planningApplicationReference: '12345/A/67890'
			});

			const validData = {
				planningApplicationReference: '12345/A/67890'
			};

			const response = await request
				.post(`${baseUrl}/${appealId}/lpa-reference/change`)
				.send(validData);

			expect(response.statusCode).toBe(302);
			expect(response.text).toBe('Found. Redirecting to /appeals-service/appeal-details/1');
		});

		it('should redirect to the appellant case page when planningApplicationReference is valid and was opened in appellant case', async () => {
			const appealId = appealData.appealId.toString();
			nock('http://test/').patch(`/appeals/${appealId}`).reply(200, {
				planningApplicationReference: '12345/A/67890'
			});

			const validData = {
				planningApplicationReference: '12345/A/67890'
			};

			const response = await request
				.post(`${baseUrl}/${appealId}/appellant-case/lpa-reference/change`)
				.send(validData);

			expect(response.statusCode).toBe(302);
			expect(response.text).toBe(
				'Found. Redirecting to /appeals-service/appeal-details/1/appellant-case'
			);
		});
	});
});
