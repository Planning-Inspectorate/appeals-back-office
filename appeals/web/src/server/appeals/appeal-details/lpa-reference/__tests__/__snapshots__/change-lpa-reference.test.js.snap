// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`change-lpa-reference GET /change should render changeLpaReference page when loaded from appeal details 1`] = `
"<main class="govuk-main-wrapper" id="main-content">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds"><span class="govuk-caption-l">Appeal 351062</span>
            <h1 class="govuk-heading-l">What is the application reference number?</h1>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full">
                    <form method="POST" novalidate="novalidate">
                        <div class="govuk-form-group">
                            <input class="govuk-input" id="planning-application-reference"
                            name="planningApplicationReference" type="text" value="48269/APP/2021/1482">
                        </div>
                        <button type="submit" data-prevent-double-click="true" class="govuk-button"
                        data-module="govuk-button">Continue</button>
                    </form>
                </div>
            </div>
        </div>
    </div>
</main>"
`;

exports[`change-lpa-reference GET /change should render changeLpaReference page when loaded from appellant case 1`] = `
"<main class="govuk-main-wrapper" id="main-content">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds"><span class="govuk-caption-l">Appeal 351062</span>
            <h1 class="govuk-heading-l">What is the application reference number?</h1>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full">
                    <form method="POST" novalidate="novalidate">
                        <div class="govuk-form-group">
                            <input class="govuk-input" id="planning-application-reference"
                            name="planningApplicationReference" type="text" value="48269/APP/2021/1482">
                        </div>
                        <button type="submit" data-prevent-double-click="true" class="govuk-button"
                        data-module="govuk-button">Continue</button>
                    </form>
                </div>
            </div>
        </div>
    </div>
</main>"
`;

exports[`change-lpa-reference POST /change should re-render changeLpaReference page with an error when planningApplicationReference is empty 1`] = `
"<main class="govuk-main-wrapper" id="main-content">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-error-summary" data-module="govuk-error-summary">
                <div role="alert">
                    <h2 class="govuk-error-summary__title"> There is a problem</h2>
                    <div class="govuk-error-summary__body">
                        <ul class="govuk-list govuk-error-summary__list">
                            <li><a href="#planning-application-reference">Enter the application reference number</a>
                            </li>
                        </ul>
                    </div>
                </div>
            </div>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds"><span class="govuk-caption-l">Appeal 351062</span>
            <h1 class="govuk-heading-l">What is the application reference number?</h1>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full">
                    <form method="POST" novalidate="novalidate">
                        <div class="govuk-form-group govuk-form-group--error">
                            <p id="planning-application-reference-error"
                            class="govuk-error-message"><span class="govuk-visually-hidden">Error:</span> Enter the LPA application
                                reference</p>
                            <input class="govuk-input govuk-input--error" id="planning-application-reference"
                            name="planningApplicationReference" type="text" value="" aria-describedby="planning-application-reference-error">
                        </div>
                        <button type="submit" data-prevent-double-click="true" class="govuk-button"
                        data-module="govuk-button">Continue</button>
                    </form>
                </div>
            </div>
        </div>
    </div>
</main>"
`;

exports[`change-lpa-reference POST /change should re-render changeLpaReference page with an error when planningApplicationReference is over 100 characters 1`] = `
"<main class="govuk-main-wrapper" id="main-content">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-error-summary" data-module="govuk-error-summary">
                <div role="alert">
                    <h2 class="govuk-error-summary__title"> There is a problem</h2>
                    <div class="govuk-error-summary__body">
                        <ul class="govuk-list govuk-error-summary__list">
                            <li><a href="#planning-application-reference">Application reference number must be 100 characters or less</a>
                            </li>
                        </ul>
                    </div>
                </div>
            </div>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds"><span class="govuk-caption-l">Appeal 351062</span>
            <h1 class="govuk-heading-l">What is the application reference number?</h1>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full">
                    <form method="POST" novalidate="novalidate">
                        <div class="govuk-form-group govuk-form-group--error">
                            <p id="planning-application-reference-error" class="govuk-error-message"><span class="govuk-visually-hidden">Error:</span> Application reference
                                number must be 100 characters or less</p>
                            <input class="govuk-input govuk-input--error"
                            id="planning-application-reference" name="planningApplicationReference"
                            type="text" value="12345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901"
                            aria-describedby="planning-application-reference-error">
                        </div>
                        <button type="submit" data-prevent-double-click="true" class="govuk-button"
                        data-module="govuk-button">Continue</button>
                    </form>
                </div>
            </div>
        </div>
    </div>
</main>"
`;
